sql/analyzer/autocommit.go

```diff
@@ -52,47 +57,46 @@ func hasShowWarningsNode(n sql.Node) bool {
 
 // GetTransactionDatabase returns the name of the database that should be considered current for the transaction about
 // to begin. The database is not guaranteed to exist.
-// For USE DATABASE statements, we consider the transaction database to be the one being USEd
-func GetTransactionDatabase(ctx *sql.Context, parsed sql.Node) string {
-	var dbName string
-	switch n := parsed.(type) {
-	case *plan.QueryProcess, *plan.TransactionCommittingNode, *plan.RowUpdateAccumulator:
-		return GetTransactionDatabase(ctx, n.(sql.UnaryNode).Child())
-	case *plan.Use, *plan.CreateProcedure, *plan.DropProcedure, *plan.CreateTrigger, *plan.DropTrigger,
-		*plan.CreateTable, *plan.InsertInto, *plan.AlterIndex, *plan.AlterAutoIncrement, *plan.AlterPK,
-		*plan.DropColumn, *plan.RenameColumn, *plan.ModifyColumn:
-		database := n.(sql.Databaser).Database()
-		if database != nil {
-			dbName = database.Name()
+// For USE DATABASE statements, we consider the transaction database to be the one being USEd.
+// If any errors are encountered determining the database for the transaction, an empty string and an error are returned.
+func GetTransactionDatabase(ctx *sql.Context, parsed sql.Node) (string, error) {
+	dbNames := make(map[string]struct{})
+	transform.Inspect(parsed, func(node sql.Node) bool {
+		switch n2 := node.(type) {
```

Comment:
Generally better to shadow the original name in type switches, unless there's some compelling reason not to (like satisfying an interface you need to return)